expression/builtin_grouping.go

```diff
@@ -112,62 +120,97 @@ func (b *BuiltinGroupingImplSig) Clone() builtinFunc {
+		for _, groupingMark := range groupingMarks {
+			if len(groupingMark) != 1 {
+				return errors.Errorf("Invalid number of groupingID. mode: %d, number of groupingID: %d", mode, len(b.groupingMarks))
+			}
+		}
 	}
 	return nil
 }
 
-func (b *BuiltinGroupingImplSig) groupingImplBitAnd(groupingID uint64, metaGroupingID uint64) int64 {
-	if groupingID&metaGroupingID > 0 {
-		return 1
+func (b *BuiltinGroupingImplSig) groupingImplBitAnd(groupingID uint64) int64 {
+	groupingMarks := b.getMetaGroupingMarks()
+	res := uint64(0)
+	for _, groupingMark := range groupingMarks {
+		// for Bit-And mode, there is only one element in groupingMark.
+		for k := range groupingMark {
+			res <<= 1
+			if groupingID&k <= 0 {
+				// col is not needed, being filled with null and grouped. = 1
+				res += 1
+			}
+			// col is needed in this grouping set, meaning not being grouped. = 0
+		}
 	}
-	return 0
+	return int64(res)
 }
 
-func (b *BuiltinGroupingImplSig) groupingImplNumericCmp(groupingID uint64, metaGroupingID uint64) int64 {
-	if groupingID > metaGroupingID {
-		return 1
+func (b *BuiltinGroupingImplSig) groupingImplNumericCmp(groupingID uint64) int64 {
+	groupingMarks := b.getMetaGroupingMarks()
+	res := uint64(0)
+	for _, groupingMark := range groupingMarks {
+		// for Num-Cmp mode, there is only one element in groupingMark.
+		for k := range groupingMark {
+			res <<= 1
+			if groupingID <= k {
+				// col is not needed, being filled with null and grouped. = 1
+				res += 1
+			}
+			// col is needed, meaning not being grouped. = 0
+		}
 	}
-	return 0
+	return int64(res)
 }
 
 func (b *BuiltinGroupingImplSig) groupingImplNumericSet(groupingID uint64) int64 {
-	groupingIDs := b.getMetaGroupingMarks()
-	_, ok := groupingIDs[groupingID]
-	if ok {
-		return 0
+	groupingMarks := b.getMetaGroupingMarks()
+	res := uint64(0)
+	for _, groupingMark := range groupingMarks {
+		res <<= 1
+		// for Num-Set mode, traverse the slice to find the match.
+		_, ok := groupingMark[groupingID]
+		if !ok {
+			// in Num-Set mode, this map maintains the needed-col's grouping set (GIDs)
+			// when ok is NOT true, col is not needed, being filled with null and grouped. = 1
+			res += 1
+		}
+		// it means col is needed, meaning not being filled with null and grouped. = 0
 	}
-	return 1
+	return int64(res)
 }
 
+// since grouping function may have multi args like grouping(a,b), so the source columns may greater than 1.
+// reference: https://dev.mysql.com/blog-archive/mysql-8-0-grouping-function/
+// Let's say GROUPING(b,a) group by a,b with rollup. (Note the b,a sequence is reversed from gby item)
+// if GROUPING (b,a) returns 3 (11 in bits), it means that NULL in column “b” and NULL in column “a” for that
+// row is produced by a ROLLUP operation. If result is 2 (10 in bits), meaning NULL in column “a” alone is the
+// result of ROLLUP operation.
+//
+// Formula: GROUPING(x,y,z) = GROUPING(x) << 2 + GROUPING(y) << 1 + GROUPING(z)
```

Comment:
```suggestion
// Formula: GROUPING(x,y,z) = (GROUPING(x) << 2) | (GROUPING(y) << 1) | GROUPING(z)
```
This is correct and more precise since `GROUPING(x)` is a single bit?

Reply:
great desc~